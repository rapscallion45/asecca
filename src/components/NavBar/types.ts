/*
 * Nav Bar type definitions
 *
 * These type definitions are used in the applications Nav Bar component.
 */

/**
 * Nav Bar sidebar open callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @callback INavBarOnSidebarOpenCallback
 * @returns {void} - no return value
 */
export interface INavBarOnSidebarOpenCallback {
  (): void;
}
