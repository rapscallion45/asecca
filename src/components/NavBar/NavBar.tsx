import { FC } from 'react';
import Link from 'next/link';
import Image from 'next/image';
import { styled } from '@mui/material/styles';
import { Box, AppBar, Toolbar, IconButton, useTheme } from '@mui/material';
import MenuIcon from '@mui/icons-material/Menu';
import { APP_BAR_MOBILE, SIDEBAR_DRAWER_WIDTH } from '@/constants/constants';
import { INavBarOnSidebarOpenCallback } from './types';

/**
 * NavBar Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef INavBarProps
 * @prop {boolean} showLogo - flag to show or hide logo
 * @prop {boolean} fullWidth - flag to show if navbar is screen fullwidth
 * @prop {INavBarOnSidebarOpenCallback} onOpenSidebar - callback handler for when sidebar is opened
 */
interface INavBarProps {
  showLogo?: boolean;
  fullWidth?: boolean;
  onOpenSidebar?: INavBarOnSidebarOpenCallback;
}

/**
 * NavBar
 *
 * Application main Navbar component
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @component
 * @param {INavBarProps} props - component props
 * @returns {FC} - navbar functional component
 */
const NavBar: FC<INavBarProps> = (props) => {
  const { fullWidth, showLogo = false, onOpenSidebar } = props;
  const theme = useTheme();

  /**
   * Nav Bar Root Style
   *
   * Application nav bar parent component styling
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.1
   *
   * @component
   * @return {Component} - styled nav bar root component
   */
  const RootStyle = styled(AppBar)(({ theme: MUITheme }) => ({
    boxShadow: 'none',
    backgroundColor:
      MUITheme.palette.mode === 'light'
        ? MUITheme.palette.grey[400]
        : MUITheme.palette.common.black,
    [MUITheme.breakpoints.up('lg')]: {
      width: fullWidth ? `100%` : `calc(100% - ${SIDEBAR_DRAWER_WIDTH + 1}px)`,
    },
  }));

  /**
   * Nav Bar Toolbar Style
   *
   * Application nav bar toolbar component styling
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.1
   *
   * @component
   * @return {Component} - styled nav bar toolbar component
   */
  const ToolbarStyle = styled(Toolbar)(({ theme: MUITheme }) => ({
    minHeight: APP_BAR_MOBILE,
    [MUITheme.breakpoints.up('lg')]: {
      padding: MUITheme.spacing(0, 5),
    },
  }));

  return (
    <RootStyle>
      <ToolbarStyle>
        {!showLogo && (
          <>
            <IconButton
              onClick={onOpenSidebar}
              sx={{ mr: 1, mt: '3px', color: 'text.primary' }}
            >
              <MenuIcon
                htmlColor={theme.palette.mode === 'dark' ? 'white' : 'black'}
              />
            </IconButton>
            <Link href="/">
              {theme.palette.mode === 'dark' && (
                <Image
                  src="/logowhite.webp"
                  alt="Asecca logo"
                  width={150}
                  height={30}
                  style={{ marginTop: '5px' }}
                />
              )}
              {theme.palette.mode === 'light' && (
                <Image
                  src="/logoblack.webp"
                  alt="Asecca logo"
                  width={150}
                  height={30}
                  style={{ marginTop: '5px' }}
                />
              )}
            </Link>
          </>
        )}
        {showLogo && (
          <Link href="/">
            {theme.palette.mode === 'dark' && (
              <Image
                src="/logowhite.webp"
                alt="Asecca logo"
                width={150}
                height={30}
                style={{ marginTop: '5px' }}
              />
            )}
            {theme.palette.mode === 'light' && (
              <Image
                src="/logoblack.webp"
                alt="Asecca logo"
                width={150}
                height={30}
                style={{ marginTop: '5px' }}
              />
            )}
          </Link>
        )}

        <Box sx={{ flexGrow: 1 }} />
      </ToolbarStyle>
    </RootStyle>
  );
};

export default NavBar;
