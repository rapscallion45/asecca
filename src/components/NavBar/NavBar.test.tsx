import React from 'react';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import NavBar from './NavBar';

/**
 * NavBar Unit Tests
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 */
describe('NavBar', () => {
  it('Renders correctly', () => {
    /** perform snapshot test */
    const tree = renderer.create(<NavBar showLogin />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('Login Button', () => {
    it('Should render if showLogin prop set', async () => {
      /** Arrange */
      const showLogin: boolean = true;

      /** Act */
      render(<NavBar showLogin={showLogin} />);

      /** Assert - login button should show */
      expect(screen.queryByText('Logout')).toBeInTheDocument();
    });

    it('Should not render if showLogin prop not set', async () => {
      /** Arrange */
      const showLogin: boolean = false;

      /** Act */
      render(<NavBar showLogin={showLogin} />);

      /** Assert - login button should show */
      expect(screen.queryByText('Logout')).toBeNull();
    });

    it('Should not render by default', async () => {
      /** Arrange */
      /** Act */
      render(<NavBar />);

      /** Assert - login button should show */
      expect(screen.queryByText('Logout')).toBeNull();
    });
  });
});
