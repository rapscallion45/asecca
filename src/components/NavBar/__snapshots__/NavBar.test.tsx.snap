// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar Renders correctly 1`] = `
<header
  className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-1xzpisq-MuiPaper-root-MuiAppBar-root"
>
  <div
    className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1lg7r32-MuiToolbar-root"
  >
    <button
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-68dm4j-MuiButtonBase-root-MuiIconButton-root"
      disabled={false}
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        color="black"
        data-testid="MenuIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
        />
      </svg>
    </button>
    <a
      href="/"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onTouchStart={[Function]}
    >
      <img
        alt="Asecca logo"
        data-nimg="1"
        decoding="async"
        height={30}
        loading="lazy"
        onError={[Function]}
        onLoad={[Function]}
        src="/_next/image?url=%2Flogoblack.webp&w=384&q=75"
        srcSet="/_next/image?url=%2Flogoblack.webp&w=256&q=75 1x, /_next/image?url=%2Flogoblack.webp&w=384&q=75 2x"
        style={
          {
            "color": "transparent",
            "marginTop": "5px",
          }
        }
        width={150}
      />
    </a>
    <div
      className="MuiBox-root css-i9gxme"
    />
  </div>
</header>
`;
